Source/ProgressionSystemRuntime/Private/Data/PSWorldSubsystem.cpp

```diff
@@ -0,0 +1,177 @@
+}
+
+// Load game from save file or create a new one (does initial load from data table)
+void UPSWorldSubsystem::LoadGameFromSave()
+{
+	// Check if the save game file exists
+	if (UGameplayStatics::DoesSaveGameExist(SaveGameInstanceInternal->GetSaveSlotName(), SaveGameInstanceInternal->GetSaveSlotIndex()))
+	{
+		SaveGameInstanceInternal = Cast<UPSSaveGameData>(UGameplayStatics::LoadGameFromSlot(SaveGameInstanceInternal->GetSaveSlotName(), SaveGameInstanceInternal->GetSaveSlotIndex()));
+	}
+	else
+	{
+		// Save file does not exist
+		// do initial load from data table
+		TMap<FName, FPSRowData> SavedProgressionRows;
+		const UDataTable* ProgressionDataTable = UPSDataAsset::Get().GetProgressionDataTable();
+		checkf(ProgressionDataTable, TEXT("ERROR: 'ProgressionDataTableInternal' is null"));
+		UMyDataTable::GetRows(*ProgressionDataTable, SavedProgressionRows);
+		SaveGameInstanceInternal = Cast<UPSSaveGameData>(UGameplayStatics::CreateSaveGameObject(UPSSaveGameData::StaticClass()));
+
+		if (SaveGameInstanceInternal)
+		{
+			SaveGameInstanceInternal->SetProgressionMap(SavedProgressionRows);
+		}
+	}
+	SetFirstElemetAsCurrent();
+}
+
+// Always set first levels as unlocked on begin play
+void UPSWorldSubsystem::SetFirstElemetAsCurrent()
+{
+	if (SaveGameInstanceInternal)
+	{
+		constexpr int32 FirstElementIndex = 0;
+		SaveGameInstanceInternal->SetCurrentProgressionRowByIndex(FirstElementIndex);
+		const FName RowName = SaveGameInstanceInternal->GetCurrentRowName();
+		SaveGameInstanceInternal->UnlockLevelByName(RowName);
+	}
+	SaveDataAsync();
+}
+
+// Saves the progression to the local files
+void UPSWorldSubsystem::SaveDataAsync()
+{
+	checkf(SaveGameInstanceInternal, TEXT("ERROR: 'SaveGameInstanceInternal' is null"));
+	UGameplayStatics::AsyncSaveGameToSlot(SaveGameInstanceInternal, SaveGameInstanceInternal->GetSaveSlotName(), SaveGameInstanceInternal->GetSaveSlotIndex());
+}
+
+// Removes all saved data of the Progression system and creates a new empty data
+void UPSWorldSubsystem::ResetSaveGameData()
+{
+	const FString& SlotName = SaveGameInstanceInternal->GetSaveSlotName();
+	const int32 UserIndex = SaveGameInstanceInternal->GetSaveSlotIndex();
+
+	UGameplayUtilsLibrary::ResetSaveGameData(SaveGameInstanceInternal, SlotName, UserIndex);
+
+	// Re-load a new save game object. Load game from save creates a save file if there is no such
+	LoadGameFromSave();
+	UPSHUDComponent* PSHUDComponent = GetProgressionSystemHUDComponent();
+	checkf(&PSHUDComponent, TEXT("ERROR: 'PSHUDComponent' is null"));
+	PSHUDComponent->UpdateProgressionWidgetForPlayer();
+}
+
+// Unlocks all levels of the Progression System
+void UPSWorldSubsystem::UnlockAllLevels()
+{
+	SaveGameInstanceInternal->UnlockAllLevels();
+	SaveDataAsync();
+	UPSHUDComponent* PSHUDComponent = GetProgressionSystemHUDComponent();
+	checkf(&PSHUDComponent, TEXT("ERROR: 'PSHUDComponent' is null"));
```

Comment:
check + address (`checkf(&`) is not valid, never do it like here and below.

Instead of using `&PSHUDComponent` (address of pointer), use `PSHUDComponent` directly in the check: `checkf(PSHUDComponent, TEXT("ERROR: 'PSHUDComponent' is null"));`

Fix it everywhere.

Why? Your `&PSHUDComponent` actually checks if the address of the pointer is valid, which it always is, since it's a local variable

Reply:
done https://github.com/h4rdmol/ProgressionSystem/pull/1/commits/c4d2490442b6a9a85b9663d2f765f78d50522f42
there is that for cases where I get address of pointer i was trying to use it. Example: 

UPSHUDComponent& PSHUDComponent = GetProgressionSystemHUDComponent();
checkf(&PSHUDComponent, TEXT("ERROR: 'PSHUDComponent' is null"));

// pretend that GetProgressionSystemHUDComponent returns address.  Since check in that do not want to check  I though this will force to read data